x-pack/plugins/lens/public/editor_frame_plugin/editor_frame.tsx

```diff
@@ -4,34 +4,139 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import React from 'react';
-import { Datasource, Visualization } from '../types';
+import React, { useState, useEffect } from 'react';
+import { Datasource, Visualization, DatasourcePublicAPI } from '../types';
 
 interface EditorFrameProps {
   datasources: { [key: string]: Datasource };
   visualizations: { [key: string]: Visualization };
+
+  activeDatasource: string | null;
+}
+
+interface DatasourceState {
+  [key: string]: {
+    state: any;
+    setState: (state: any) => void;
+    api: DatasourcePublicAPI;
```

Comment:
We shouldn't store these in the state as they contain functions which is a code smell in general plus it doesn't help us because the datasource can be accessed equally easily by `props.datasources[key]`. The api object has to be re-created each time the state is updated which means storing it in the state just brings in the risk of forgetting that and having a stale version of the api. As it is "derived state" (https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization) it should be created as part of the render function and, if we are concerned about performance, be memoized.